global/scripts/debug.shlib

```diff
@@ -0,0 +1,328 @@
+#   
+#   To disable error trapping altogether run `debug_disable_trap`
+#      (Note that if `set -e` is set in the calling script, the script would still exit upon error 
+#       through that mechanism).
+#   To disable exit upon trapped error run `debug_disable_exit`
+#       In this case (trap doesn't exit), execution proceeds to the next line of the script, 
+#       with `$?` still set to whatever value the failed command errored with.
+#
+# Input:
+#
+#   $1 - line number
+#   $2 - command ran
+#   $3 - exit status
+
+debug_on_error()
+{    
+    if [ ! $HCP_DEBUG_USETRAP == 'OFF' ]; then        
+
+        local me=`basename "$0"`
+        local startreport=""
+        local endreport=""
+        local command=$(eval debug_first "$2")
+
+        # --- output a regular HCP error log
+
+        log_Err "'$command' command failed with return code: $3"
+
+        # --- are we printing in color
+
+        if [ "$HCP_DEBUG_COLOR" == "TRUE" ] ; then
+           startreport="\033[33m"
+           endreport="\033[0m"
+        fi
+        
+        # --- output a detailed report if usetrap is not turned off    
+        echo                                                               
+        echo -e "${startreport}===> ERROR: Command returned with nonzero exit code"
+        echo "---------------------------------------------------"
+        echo "         script: $me"                                            
+        echo "stopped at line: $1"                                    
+        echo "           call: $2"
+        echo "  expanded call: $(eval echo "$2")"                            
+        echo "      exit code: $3"                                        
+        echo -e "---------------------------------------------------${endreport}" 
+        echo                                                               
+        
+        # --- check if ok to exit        
+        if [ ! $HCP_DEBUG_EXIT_ON_ERROR == 'OFF' ]; then
+            if [[ "$-" == *i* ]]
+            then
+                echo
+                echo -n "execution paused due to error, press enter to continue or ctrl+c to halt ..."
+                read
+            else
+                debug_red_echo "===> Aborting execution!"
+                echo ""
+                exit $3
+            fi
+        fi
+    fi
+}
+
+function debug_first()
+{
+    echo "$1"
+}
+
+debug_defaults()
+{
+    debug_enable_trap
+    debug_enable_exit
+    #the ${<var>:-} pattern exists to work properly under set -eu, mac's old bash doesn't have [[ -v <var> ]]
+    if [[ "${HCP_DEBUG_COLOR:-}" == "" ]]
+    then
+        debug_disable_color
+    fi
+    if [[ "${HCP_VERBOSE:-}" == "" ]]
+    then
+        debug_disable_verbose
+    fi
+}
+
+# == Logging functionality
+
+if [[ "$(type -t log_Err)" != "function" ]]
+then
+    source "${HCPPIPEDIR}/global/scripts/debug.shlib" "$@" # Debugging functions; also sources log.shlib
```

Comment:
Looks like an over-aggressive find-replace.  Aren't we supposed to source log.shlib here?

Reply:
It was not find-replace. As per @coalsont instructions I replaced all log.shlib sourcing with debugging ones.

Reply:
Well, I  believe that this is the one instance of `source log.shlib` that we need to keep (in `debug.shlib` itself).  Otherwise, we have an infinite loop of `debug.shlib` continually sourcing itself.

Reply:
:-) I completely overlooked that it was in the debug.shlib itself. Fixed.